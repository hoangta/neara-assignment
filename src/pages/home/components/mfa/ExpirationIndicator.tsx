import { useEffect, useState, useRef } from 'react'
import { observer } from 'mobx-react'
import styles from './ExpirationInficator.module.css'

type Props = {
  exp: Date
}

const indicatorRadius = 14
const strokeDasharray = 2 * indicatorRadius * Math.PI

const ExpirationIndicator = ({ exp }: Props) => {
  const circleRef = useRef<SVGCircleElement>(null)
  const [remainingSeconds, setRemainingSeconds] = useState<number>(0)

  useEffect(() => {
    const remainingSeconds = (exp.getTime() - new Date().getTime()) / 1000
    circleRef.current?.animate([{ strokeDashoffset: 0 }, { strokeDashoffset: strokeDasharray }], {
      duration: remainingSeconds * 1000,
    })
    setRemainingSeconds(remainingSeconds)
    const timer = setInterval(() => setRemainingSeconds((r) => Math.max(r - 1, 0)), 1000)
    return () => clearInterval(timer)
  }, [exp])

  return (
    <div className={styles.indicatorContainer}>
      <svg className={styles.indicatorItem}>
        <circle
          ref={circleRef}
          className={styles.circle}
          cx="20"
          cy="20"
          r={indicatorRadius}
          fill="none"
          stroke="gray"
          strokeWidth="2"
          strokeDasharray={strokeDasharray}
          strokeDashoffset={strokeDasharray}
        />
      </svg>
      <div className={styles.countdownLabel}>{remainingSeconds.toFixed()}</div>
    </div>
  )
}

export default observer(ExpirationIndicator)
