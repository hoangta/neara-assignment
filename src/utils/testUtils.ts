import { MFA } from "../types/mfa";

export const mockStore = () => {
  jest.mock("../stores/mfaStore", () =>
    jest.fn().mockImplementation(() => ({
      mfas: mockMFAs,
      createMFA: jest.fn(),
      switchMFAs: jest.fn(),
    }))
  );
};

export const mockSystemTime = (date: Date) => {
  jest.useFakeTimers("modern");
  jest.setSystemTime(date);
};

export const mockSVGElementAnimate = () => {
  const circleAnimateSpy = jest.fn();
  SVGElement.prototype.animate = circleAnimateSpy;
  return circleAnimateSpy;
};

export const mockMFAs: MFA[] = [
  {
    provider: "Test Provider",
    code: "012345",
    image: "www.testurl.com",
    exp: new Date(2023, 2, 2),
  },
  {
    provider: "Test Provider 2",
    code: "543210",
    image: "www.testurl.com",
    exp: new Date(2023, 2, 3),
  },
];
