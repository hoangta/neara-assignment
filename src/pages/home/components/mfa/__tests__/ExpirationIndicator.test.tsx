import React from "react";
import { render, act } from '@testing-library/react';
import ExpirationIndicator from '../ExpirationIndicator'
import { mockSVGElementAnimate } from "../../../../../utils/testUtils";

const circleAnimateSpy = mockSVGElementAnimate();

describe('Pages/Home/Components/MFA/ExpirationIndicator', () => {
    const testExp = new Date(2023, 2, 2)

    it('should render correctly', () => {
        const { asFragment } = render(<ExpirationIndicator exp={testExp} />);
        expect(asFragment()).toMatchSnapshot();
    })

    it('should render remaining seconds correctly', () => {
        const mockRemainingSeconds = 86400
        const { getByText } = render(<ExpirationIndicator exp={testExp} />);
        expect(circleAnimateSpy).toBeCalledWith([{ strokeDashoffset: 0 }, { strokeDashoffset: 2 * 14 * Math.PI }], {
            duration: mockRemainingSeconds * 1000
        });
        expect(getByText(mockRemainingSeconds.toFixed())).toBeInTheDocument();
        const secsToPass = 2;
        act(() => {
            jest.advanceTimersByTime(secsToPass * 1000);
        })
        expect(getByText((mockRemainingSeconds - secsToPass).toFixed())).toBeInTheDocument();
    })
})