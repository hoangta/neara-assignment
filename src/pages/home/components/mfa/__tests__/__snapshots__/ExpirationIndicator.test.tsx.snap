// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages/Home/Components/MFA/ExpirationIndicator should render correctly 1`] = `
<DocumentFragment>
  <div
    class="indicatorContainer"
  >
    <svg
      class="indicatorItem"
    >
      <circle
        class="circle"
        cx="20"
        cy="20"
        fill="none"
        r="14"
        stroke="gray"
        stroke-dasharray="87.96459430051421"
        stroke-dashoffset="87.96459430051421"
        stroke-width="2"
      />
    </svg>
    <div
      class="countdownLabel"
    >
      86400
    </div>
  </div>
</DocumentFragment>
`;
